File: 0046-permutations/0046-permutations.java
class Solution {
        //optimez the space about freq array 2 approach
        private static void f(int ind,int []arr,List<List<Integer>>ans){
                if(ind==arr.length){List<Integer>ds=new ArrayList<>();
                        for(int i:arr){
                                
                                ds.add(i);
                        }
                        ans.add(new ArrayList<>(ds));
                        return;
                }
                for(int i=ind;i<arr.length;i++){
                        swap(i,ind,arr);
                        f(ind+1,arr,ans);
                        swap(i,ind,arr);
                }
        }
                private static void swap(int i ,int ind,int []arr){
                        int t=arr[i];
                        arr[i]=arr[ind];
                        arr[ind]=t;
                }
        
        public List<List<Integer>> permute(int[] arr) {
                List<List<Integer>>ans=new ArrayList<>();
                f(0,arr,ans);
                return ans;
        }
}
        
        
        //1
 /*       private static void f(int []nums,boolean[]freq,List<List<Integer>>ans,List<Integer>ds){
                if(ds.size()==nums.length){
                        ans.add(new ArrayList<>(ds));
                        return;
                }
                for(int i=0;i<nums.length;i++)
                {
                        if(!freq[i])
                        {
                                freq[i]=true;
                                ds.add(nums[i]);
                                f(nums,freq,ans,ds);
                                ds.remove(ds.size()-1);
                                freq[i]=false;
                        }
                }
                
        }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>>ans=new ArrayList<>();
            boolean[]freq=new boolean[nums.length];
            f(nums,freq,ans,new ArrayList<>());
            return ans;            
    }
}*/

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void p(int ind,vector<int>&nums,vector<vector<int>>&ans){
        if(ind==nums.size()){
            ans.push_back(nums);
            return;
        }
        for(int i=ind;i<nums.size();i++){
            swap(nums[i],nums[ind]);
            p(ind+1,nums,ans);
            swap(nums[i],nums[ind]);
        }
    }
    
    // void p(vector<int>nums,vector<vector<int>>&ans,vector<int>&ds,vector<int>&freq){
    //     if(ds.size()==nums.size()){
    //         ans.push_back(ds);
    //         return;
    //     }
    //     for(int i=0;i<nums.size();i++){
    //         if(freq[i]==false){
    //             freq[i]=true;
    //             ds.push_back(nums[i]);
    //             p(nums,ans,ds,freq);
    //             ds.pop_back();
    //             freq[i]=false;
    //         }
    //     }
    // }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>>ans;
        
        p(0,nums,ans);
        return ans;
    }
};